llvm/lib/Target/Mips/NanoMipsLoadStoreMultiple.cpp

```diff
@@ -0,0 +1,359 @@
+      computeLiveIns(LiveRegs, *Prev.MBB);
+      assert(Register::isPhysicalRegister(DesiredRtNo) &&
+             "Desired register is not physical!");
+      if (!LiveRegs.available(*MRI, (DesiredRtReg)))
+        return false;
+
+      RegGap = true;
+      return true;
+    } else {
+      return true;
+    }
+  } else {
+    // "full" GAP
+    // lw a0, 8(a4)
+    // lw a1, 12(a4)
+    // lw a3, 20(a4)
+    bool OffsetOk = ((Next.Offset - Prev.Offset) % 4) == 0;
+    int Gap = (Next.Offset - Prev.Offset) / 4 - 1;
+    if (OffsetOk && (CurrSeqSize + Gap + 1 <= 8) &&
+        Next.Rt == RC.getRegister(PrevRtNo + Gap + 1)) {
+      LivePhysRegs LiveRegs(*TRI);
+      computeLiveIns(LiveRegs, *Prev.MBB);
+      for (size_t i = 0; i < Gap; i++) {
+        assert(Register::isPhysicalRegister(DesiredRtNo + i) &&
+               "Desired register is not physical!");
+        if (!LiveRegs.available(*MRI, (DesiredRtReg)))
+          return false;
+        DesiredRtReg = RC.getRegister(DesiredRtNo + i + 1);
+      }
+      GapSize += Gap;
+      CurrSeqSize += Gap;
+      return true;
+    }
+  }
+  return false;
+}
+
+bool NMLoadStoreMultipleOpt::generateLoadStoreMultiple(MachineBasicBlock &MBB,
+                                                       bool IsLoad) {
+  bool Modified = false;
+  struct Candidate {
+    InstrList Sequence;
+    size_t GapSize;
+    bool Move = false;
+  };
+  InstrList SequenceToSort;
+  SmallVector<InstrList, 3> SequenceList;
+  for (auto &MI : MBB) {
+    // CFI and debug instructions don't break the sequence.
+    if (MI.isCFIInstruction() || MI.isDebugInstr())
+      continue;
+    if (isValidLoadStore(MI, IsLoad, SequenceToSort)) {
+      SequenceToSort.push_back(&MI);
+      continue;
+    }
+    if (SequenceToSort.size() > 1) {
+      SequenceList.push_back(SequenceToSort);
+      SequenceToSort.clear();
+    }
+  }
+
+  SmallVector<InstrList, 3> Candidates;
+  InstrList Sequence;
+  size_t GapSize = 0;
+  size_t SeqSize = 0;
+  bool RegGap = false;
+  for (size_t i = 0; i < SequenceList.size(); i++) {
+    sortLoadStoreList(SequenceList[i], IsLoad);
+    for (auto &MI : SequenceList[i]) {
+      // Sequences cannot be longer than 8 instructions.
+      if (SeqSize == 8) {
+        Candidates.push_back({Sequence, GapSize});
+        Sequence.clear();
+        GapSize = 0;
+        SeqSize = 0;
+        RegGap = false;
+      }
+      // When starting a new sequence, there's no need to do any checks.
+      if (Sequence.empty()) {
+        Sequence.push_back(MI);
+        SeqSize = 1;
+        continue;
+      }
+
+      if (!isValidNextLoadStore(Sequence.back(), MI, GapSize, SeqSize,
+                                RegGap)) {
+        if (SeqSize > 1)
+          Candidates.push_back({Sequence, GapSize});
+        Sequence.clear();
+        GapSize = 0;
+        SeqSize = 0;
+        RegGap = false;
+      }
+
+      Sequence.push_back(MI);
+      SeqSize++;
+
+      if (RegGap) {
```

Comment:
I agree that allowing multiple gaps could be beneficial if register liveness allows it, as it might help optimize size further. However, I would say that incorporating this feature into the current implementation is not trivial and would require quite a few changes. For now, our approach is to test and evaluate the benefits of the current implementation. Based on the results and the impact it has, we can consider extending support for multiple gaps in the future.